### ISSUE TITLE
Grep match ending in last column losing a character

### ISSUE THREAD

**Issue by chrisduerr** [alacritty/alacritty#5397]
I'm not entirely certain what's going on here, but I noticed this while looking through some logs and was initially afraid our resize logic might have a problem. But it seems like grep has an issue where a cell is lost when the highlight ends in the last column.

Generally I'd just say this is clearly a grep bug, since XTerm/URxvt also lose a character here. But what I'm surprised by is that this does not happen in Kitty. So I'd be curious what's different about Kitty here that this doesn't occur and if Alacritty can maybe emulate that behavior.

```
XTerm:     https://paste.rs/x08.png
URxvt:     https://paste.rs/Xfy.png
Kitty:     https://paste.rs/gx1.png
Alacritty: https://paste.rs/ZgV.png

Alacritty log:
    [2021-08-05 00:27:44.695975791] [TRACE] [alacritty_terminal] Setting attribute: Bold
    [2021-08-05 00:27:44.695984440] [TRACE] [alacritty_terminal] Setting attribute: Foreground(Named(Red))
    [2021-08-05 00:27:44.695995019] [TRACE] [alacritty_terminal] Clearing line: Right
    [2021-08-05 00:27:44.696011358] [TRACE] [alacritty_terminal] Setting attribute: Reset
    [2021-08-05 00:27:44.696020507] [TRACE] [alacritty_terminal] Clearing line: Right
    [2021-08-05 00:27:44.696028046] [TRACE] [alacritty_terminal] Wrapping input
    [2021-08-05 00:27:44.696036556] [TRACE] [alacritty_terminal] Carriage return
    [2021-08-05 00:27:44.696043545] [TRACE] [alacritty_terminal] Linefeed
    [2021-08-05 00:27:44.696067823] [TRACE] [alacritty_terminal] Setting mode: BracketedPaste
```

**Comment by dnkl** [comment]
It _is_ a `grep` bug, and an old one at that: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=15444 (though https://bugzilla.redhat.com/show_bug.cgi?id=1006310 is a more interesting read).

TL;DR wokaround is to set `GREP_COLORS=ne`

**Comment by chrisduerr** [comment]
@dnkl Yeah, my assumption was that this issue exists because the cursor is in the position "after" the last column, which of course doesn't exist and usually still counts as the last column.

I'm mostly curious if there's any evidence of Kitty's behavior being right or wrong. I'd assume they just differentiate between being in the last column and having just written to the last column (thus requiring a wrap).

Are you aware of any spec explicitly pointing this out as incorrect, or software actually running into problems with Kitty's approach?

**Comment by dnkl** [comment]
> Are you aware of any spec explicitly pointing this out as incorrect, 

https://github.com/mattiase/wraptest sums up the relevant specs. In this case, I'm assuming kitty does **not** clear the LCF (last column flag) when it receives EL (erase line).

> or software actually running into problems with Kitty's approach?

No, but given this https://github.com/mattiase/wraptest/blob/master/results.txt (roughly 50% of the terminals listed there clears LCF on EL, the other 50% does not), I'd don't expect any issues with kitty's approach.

**Comment by chrisduerr** [comment]
> roughly 50% of the terminals listed there clears LCF on EL, the other 50% does not

Beautiful. Such consistent inconsistency.

**Comment by chrisduerr** [comment]
It's pretty clear that `grep` is to blame here, which is why I'm inclined to not make any exceptions even though it is an application that affects a lot of users.

**Comment by jerch** [comment]
> It's pretty clear that grep is to blame here...

I dont think it is that easy. To me it seems more like a "definition gap" for the last char in a line.

**Comment by j4james** [comment]
Note that there are two issues at play here. The first one is that an `EL` at the end of the line is going to erase the last character. The second one is that `EL` at the end of the line will reset the last column flag, so the last character will be overwritten by whatever comes next.

And as far as I can tell, Alacritty isn't actually resetting the last column flag, so it's likely just the `EL` erasure that's the problem. You can see this in the test case from the grep bug report. On XTerm the last character is replaced with `x` (this is the result of the flag being reset), while with Alacritty it's replaced with a space (this is the `EL` erasure).

So for a terminal to work with grep as it is, they would need to ignore two separate standards. Some might be willing to do that, but I doubt XTerm could ever be persuaded to go that route.

And if BCE is really the issue here, it'd probably make more sense to push terminals to support `DECECM`, so application that don't want that effect can simply turn it off. That at least doesn't require anyone to change their behaviour in a backwards incompatible way.

Just my 2c.

**Comment by chrisduerr** [comment]
> And as far as I can tell, Alacritty isn't actually resetting the last column flag, so it's likely just the EL erasure that's the problem. You can see this in the test case from the grep bug report. On XTerm the last character is replaced with x (this is the result of the flag being reset), while with Alacritty it's replaced with a space (this is the EL erasure).

That's correct, Alacritty's behavior here matches URxvt more than XTerm.

> So for a terminal to work with grep as it is, they would need to ignore two separate standards. Some might be willing to do that, but I doubt XTerm could ever be persuaded to go that route.

Exactly this is the problem. It's not about *interpreting* standards in one way or another. It's about explicitly **not** doing things correctly to work around what I'd say is a clear bug in one application. I'm pretty sure there are several places in Alacritty where I've decided to explicitly deviate in favor of more usable behavior, but I don't think this is one of those places.
